### PR TITLE
[MI-2289] Added handling of refresh token expiry and other client errors

### PR DESCRIPTION
**Pull request by manojmalik20**
Created a util function handleClientError used by both APIs and slash commands
Changed the APIs and slash commands to use the new function
Created some new constants

### CODE REVIEW

File: server/plugin/util.go [hunk 1]
```diff
@@ -193,3 +194,75 @@ func filterSubscriptionsOnRecordData(subscripitons []*serializer.SubscriptionRes
 
 	return subscripitons[:n]
 }
+
+func (p *Plugin) handleClientError(err error, w http.ResponseWriter, r *http.Request, userID string, isSysAdmin bool, statusCode int, response string) string {
```
Comment (Kshitij-Katiyar) [inline server/plugin/util.go:198]: I think its better to put 2 string arguments together and put the error at end
```suggestion
func (p *Plugin) handleClientError(w http.ResponseWriter, r *http.Request, userID, response string, isSysAdmin bool, statusCode int, err error) string {  
```

File: server/plugin/util.go [hunk 2]
```diff
@@ -193,3 +194,75 @@ func filterSubscriptionsOnRecordData(subscripitons []*serializer.SubscriptionRes
...
+			p.API.LogError(disconnectErr.Error())
+			if w != nil {
+				p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusInternalServerError, Message: genericErrorMessage})
+			}
+			return genericErrorMessage
+		}
+
+		if w != nil {
+			p.handleAPIError(w, &serializer.APIErrorResponse{ID: constants.APIErrorIDRefreshTokenExpired, StatusCode: http.StatusBadRequest, Message: constants.APIErrorRefreshTokenExpired})
+			return message
```
Comment (Kshitij-Katiyar) [inline server/plugin/util.go:215]: in this case we are returning an empty string as message haven't been updated is this correct ?

File: server/plugin/util.go [hunk 3]
```diff
@@ -193,3 +194,75 @@ func filterSubscriptionsOnRecordData(subscripitons []*serializer.SubscriptionRes
...
+			message = subscriptionsNotConfiguredErrorForAdmin
+		}
+
+		return message
+	}
+
+	if strings.EqualFold(err.Error(), constants.APIErrorIDSubscriptionsNotAuthorized) {
+		if w != nil {
+			p.handleAPIError(w, &serializer.APIErrorResponse{ID: constants.APIErrorIDSubscriptionsNotAuthorized, StatusCode: http.StatusUnauthorized, Message: constants.APIErrorSubscriptionsNotAuthorized})
+			return message
```
Comment (Kshitij-Katiyar) [inline server/plugin/util.go:238]: here as well

File: server/plugin/util.go [hunk 4]
```diff
@@ -193,3 +194,75 @@ func filterSubscriptionsOnRecordData(subscripitons []*serializer.SubscriptionRes
...
+	if strings.EqualFold(err.Error(), constants.APIErrorIDLatestUpdateSetNotUploaded) {
+		if w != nil {
+			p.handleAPIError(w, &serializer.APIErrorResponse{ID: constants.APIErrorIDLatestUpdateSetNotUploaded, StatusCode: http.StatusBadRequest, Message: constants.APIErrorLatestUpdateSetNotUploaded})
+		}
+
+		return constants.APIErrorIDLatestUpdateSetNotUploaded
+	}
+
+	if w != nil {
+		if statusCode == 0 {
```
Comment (Kshitij-Katiyar) [inline server/plugin/util.go:258]: I think for some function calls we are not sending or either sending 0 as status code and for all those cases we are returning an internal server error then we can directly send 500  as status code from all those functions calls

File: server/plugin/command.go [hunk 5]
```diff
@@ -48,14 +47,14 @@ After that, this user will have the permission to add or manage subscriptions fr
...
 	disconnectSuccessMessage                = "Disconnected your ServiceNow account."
-	listSubscriptionsErrorMessage           = "Something went wrong. Not able to list subscriptions. Check server logs for errors."
 	listSubscriptionsWaitMessage            = "Your subscriptions will be listed soon. Please wait."
 	genericWaitMessage                      = "Your request is being processed. Please wait."
 	deleteSubscriptionErrorMessage          = "Something went wrong. Not able to delete subscription. Check server logs for errors."
 	deleteSubscriptionSuccessMessage        = "Subscription successfully deleted."
 	genericErrorMessage                     = "Something went wrong."
 	invalidSubscriptionIDMessage            = "Invalid subscription ID."
 	notConnectedMessage                     = "You are not connected to ServiceNow.\n[Click here to link your ServiceNow account.](%s%s)"
+	tokenExpiredReconnectMessage            = "Your connection with ServiceNow has expired, so you need to reconnect your account.\n[Click here to link your ServiceNow account.](%s%s)"
```
Comment (ayusht2810) [inline server/plugin/command.go:57]: same

File: server/plugin/util.go (hunk 2 above)
Comment (manojmalik20) [reply server/plugin/util.go:215]: Correct.

File: server/plugin/util.go (hunk 3 above)
Comment (manojmalik20) [reply server/plugin/util.go:238]: Yes

File: server/plugin/util.go (hunk 4 above)
Comment (manojmalik20) [reply server/plugin/util.go:258]: Yes, we can do that, but that will require more effort than this. This util is created for minimizing the caller's effort.

File: server/plugin/command.go (hunk 5 above)
Comment (manojmalik20) [reply server/plugin/command.go:57]: Already updated in PR #99

File: server/plugin/util.go (hunk 4 above)
Comment (Kshitij-Katiyar) [reply server/plugin/util.go:258]: ohh ok for me it looked weird equating the status code to 0 that's why

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/util.go (hunk 4 above)
Comment (manojmalik20) [reply server/plugin/util.go:258]: That is the default value for integer type, that's why we are passing 0.

**Review by avas27JTG — APPROVED** [review]

**Pull request merged**